tfhe/src/integer/server_key/radix_parallel/sub.rs

```diff
@@ -407,6 +408,303 @@ impl ServerKey {
+            sks: &ServerKey,
+            lhs: &mut Ciphertext,
+            rhs: &Ciphertext,
+        ) -> Ciphertext {
+            sks.key.unchecked_add_assign(lhs, rhs);
+            let (carry, message) = rayon::join(
+                || sks.key.carry_extract(lhs),
+                || sks.key.message_extract(lhs),
+            );
+
+            *lhs = message;
+
+            carry
+        }
+
+        // 2_2, 3_3, 4_4
+        // If we have at least 2 bits and at least as much carries
+        if self.key.message_modulus.0 >= 4 && self.key.carry_modulus.0 >= self.key.message_modulus.0
+        {
+            self.unchecked_sub_assign(&mut result, rhs);
+
+            let mut input_carry = self.key.create_trivial(0);
+
+            // For the first block do the first step of overflow computation in parallel
+            let (_, last_block_inner_propagation) = rayon::join(
+                || {
+                    input_carry =
+                        block_add_assign_returning_carry(self, &mut result.blocks[0], &input_carry);
+                },
+                || {
+                    self.generate_last_block_inner_propagation(
+                        &lhs.blocks[num_blocks - 1],
+                        &rhs.blocks[num_blocks - 1],
+                    )
+                },
+            );
+
+            for block in result.blocks[1..num_blocks - 1].iter_mut() {
+                input_carry = block_add_assign_returning_carry(self, block, &input_carry);
+            }
+
+            // Treat the last block separately to handle last step of overflow behavior
+            let output_carry = block_add_assign_returning_carry(
+                self,
+                &mut result.blocks[num_blocks - 1],
+                &input_carry,
+            );
+            let overflowed = self.resolve_signed_overflow(
+                last_block_inner_propagation,
+                &input_carry,
+                &output_carry,
+            );
+
+            return (result, overflowed);
+        }
+
+        // 1_X parameters
+        //
+        // Same idea as other algorithms, however since we have 1 bit per block
+        // we do not have to resolve any inner propagation but it adds one more
+        // sequential PBS
+        if self.key.message_modulus.0 == 2 {
+            self.unchecked_sub_assign(&mut result, rhs);
+
+            let mut input_carry = self.key.create_trivial(0);
+            for block in result.blocks[..num_blocks - 1].iter_mut() {
+                input_carry = block_add_assign_returning_carry(self, block, &input_carry);
+            }
+
+            let output_carry = block_add_assign_returning_carry(
+                self,
+                &mut result.blocks[num_blocks - 1],
+                &input_carry,
+            );
+
+            // Encode the rule
+            // "Overflow occurred if the two inputs add the same sign and output has a different
+            // sign" using 2 PBS
```

Comment:
I'm not a 2s complement expert like you are, what you are saying is that when seeing a - b as a + (-b), if a is > 0 e.g. a = 4 and -b is also positive so let's say b = 5

a + (-b) = 4 - 5 = -1 then this is an overflow ? I'm positive I'm wrong but this is how I understand the sentence, if you have a simple formula to explain that sentence I would put it in the comment :)

Reply:
Well, the comment is not up to date with the code, the code 'encodes' a different rule for overflow, its the one I first used

- The comment says `Overflow occurred if the two inputs had the same sign and output has a different`

`4 + (-5)` -> 4 is positive, -5 is negative, the two input signs are different, this can never overflow
`4 + 5` -> both are positive (same sign), assuming 4 bits twos complement number, the result is `1001` = -7 -> different sign so overflow occured
`-3 + -3` -> both are negative , assuming 4 bits twos complement number, the result is  -6 -> same sign no overflow occured.

The code now uses a different rule that is, 'if the carry onto the last pair of bit is different that the output carry of said pair of bits, then overflow occurent`

Reply:
ok then let's update the wording in that case, I don't think I understand the new rule which seems more complex 😅 ransom of performance I guess !

Reply:
Yeah the first rules needs that we have done carry_propgation + carry-add + message_extract.

Whereas the rule on input/output carry of last bit allows to compute overflow flag while we do the carry-add+message_extract

Reply:
If you do 
```
 4: 0100
+5: 0101
----------
(-7) 1001
```

The last pair of bit, (`0`, `0`), receives as input the carry of `1 + 1`, which is `1`, so, `1 + 0 +0` emits 0 as output carry.

-> Input 1, output 0 => overflow

Reply:
Yep ok, still boggles the mind